cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -16,10 +16,192 @@
+      md5_hash_step(hash_state, hash_constants, shift_constants);
+
+      while (len > 64 + copylen) {
+        thrust::copy_n(thrust::seq, data + copylen, 64, hash_state->buffer);
+        md5_hash_step(hash_state, hash_constants, shift_constants);
+        copylen += 64;
+      }
+
+      thrust::copy_n(thrust::seq, data + copylen, len - copylen, hash_state->buffer);
+      hash_state->buffer_length = len - copylen;
+    }
+  }
+
+  template <typename T, typename std::enable_if_t<std::is_floating_point<T>::value>* = nullptr>
+  void __device__ operator()(T const& key,
+                             md5_intermediate_data* hash_state,
+                             md5_hash_constants_type const* hash_constants,
+                             md5_shift_constants_type const* shift_constants) const
+  {
+    if (isnan(key)) {
+      T nan = std::numeric_limits<T>::quiet_NaN();
+      process(nan, sizeof(T), hash_state, hash_constants, shift_constants);
+    } else if (key == T{0.0}) {
+      process(T{0.0}, sizeof(T), hash_state, hash_constants, shift_constants);
+    } else {
+      process(key, sizeof(T), hash_state, hash_constants, shift_constants);
+    }
+  }
+
+  template <typename T,
+            typename std::enable_if_t<std::is_same<T, cudf::string_view>::value>* = nullptr>
+  void __device__ operator()(T const& key,
+                             md5_intermediate_data* hash_state,
+                             md5_hash_constants_type const* hash_constants,
+                             md5_shift_constants_type const* shift_constants) const
+  {
+    uint32_t const len  = static_cast<uint32_t>(key.size_bytes());
+    uint8_t const* data = reinterpret_cast<uint8_t const*>(key.data());
+
+    hash_state->message_length += len;
+
+    if (hash_state->buffer_length + len < 64) {
+      thrust::copy_n(thrust::seq, data, len, hash_state->buffer + hash_state->buffer_length);
+      hash_state->buffer_length += len;
+    } else {
+      uint32_t copylen = 64 - hash_state->buffer_length;
+      thrust::copy_n(thrust::seq, data, copylen, hash_state->buffer + hash_state->buffer_length);
+      md5_hash_step(hash_state, hash_constants, shift_constants);
+
+      while (len > 64 + copylen) {
+        thrust::copy_n(thrust::seq, data + copylen, 64, hash_state->buffer);
+        md5_hash_step(hash_state, hash_constants, shift_constants);
+        copylen += 64;
+      }
+
+      thrust::copy_n(thrust::seq, data + copylen, len - copylen, hash_state->buffer);
+      hash_state->buffer_length = len - copylen;
+    }
+  }
+
+  template <typename T, typename std::enable_if_t<is_chrono<T>()>* = nullptr>
+  void __device__ operator()(T const& key,
+                             md5_intermediate_data* hash_state,
+                             md5_hash_constants_type const* hash_constants,
+                             md5_shift_constants_type const* shift_constants) const
+  {
+    release_assert(false && "Unsupported hash type");
```

Comment:
It doesn't look like you do any error checking on the host to verify that the column types in the input table can be hashed. If you hit this assert, it's an unrecoverable error that kills the GPU's context.

Reply:
Is there a better way to do the column type checking than what I implemented in cpp/src/hash/hashing.cu using CUDF_EXPECTS?

Reply:
Yeah, I missed it. Though it would be nicer to use an `all_of` or `any_of` instead of a `for_each`.